homeassistant/components/mqtt/__init__.py

```diff
@@ -526,36 +520,54 @@ def stop():
         return self.hass.async_add_job(stop)
 
     @asyncio.coroutine
-    def async_subscribe(self, topic, qos):
-        """Subscribe to a topic.
+    def async_subscribe(self, topic, msg_callback, qos, encoding):
+        """Set up a subscription to a topic with the provided qos.
 
         This method is a coroutine.
         """
         if not isinstance(topic, str):
-            raise HomeAssistantError("topic need to be a string!")
+            raise HomeAssistantError("topic needs to be a string!")
 
-        with (yield from self._paho_lock):
-            if topic in self.subscribed_topics:
+        subscription = Subscription(topic, msg_callback, qos, encoding)
+        self.subscriptions.append(subscription)
+
+        yield from self._async_perform_subscription(topic, qos)
+
+        @callback
+        def async_remove():
+            """Remove subscription."""
+            if subscription not in self.subscriptions:
+                raise HomeAssistantError("Can't remove subscription twice")
```

Comment:
As this line of code should only ever be executed when `async_remove` is called twice, I think it's good to really fail and not just put out an error log message.